File: tools/binary-diff-summary.py
#!/usr/bin/env python3
"""High-level change summary between two PE binary versions (patch-day prep).

Answers one question before you commit an afternoon to a game update: *how
much did this binary actually change?* This is not a byte-by-byte diff —
Diaphora and radiff2 already do function-level matching. This tool hashes
fixed-size blocks of each named section and uses multiset comparison to report
the share of content that survived the patch, was replaced, was added, or was
removed. From that it classifies the .text section as a RECOMPILE (sigs likely
survive, offsets shifted), a REFACTOR (some sigs survive), or a MAJOR_CHANGE
(assume full re-RE), and recommends whether the patch-day playbook applies.

Run it first; run tools/offset-diff.py second to confirm the specific sigs.

Usage:
    python3 binary-diff-summary.py --old old.exe --new new.exe
    python3 binary-diff-summary.py --old old.exe --new new.exe --sections .text,.rdata
    python3 binary-diff-summary.py --old old.exe --new new.exe --block-size 4096
    python3 binary-diff-summary.py --old old.exe --new new.exe --json

Block comparison is content-based, not positional: a recompile shifts every
block's file offset but leaves the block bytes intact, so positional diffing
would falsely report 100% changed. Matching by hash reveals the survivors.
"""
import sys, struct, os, json, argparse, hashlib
from collections import Counter

# .text classifier thresholds (block-level identical %, raw-size delta %).
RECOMPILE_MIN_SAME = 70.0
RECOMPILE_MAX_DELTA = 1.0
REFACTOR_MIN_SAME = 30.0
MAJOR_MAX_DELTA = 10.0


def read_u16(d, o): return struct.unpack_from('<H', d, o)[0]
def read_u32(d, o): return struct.unpack_from('<I', d, o)[0]


def get_sections(data):
    """Minimal PE parser — returns list of (name, offset, size) tuples."""
    if data[:2] != b'MZ':
        return None
    pe_off = read_u32(data, 0x3C)
    if data[pe_off:pe_off+4] != b'PE\x00\x00':
        return None
    coff = pe_off + 4
    num = read_u16(data, coff + 2)
    opt_size = read_u16(data, coff + 16)
    sec_off = coff + 20 + opt_size
    sections = []
    for i in range(num):
        s = sec_off + i * 40
        name = data[s:s+8].rstrip(b'\x00').decode('ascii', errors='replace')
        rsize = read_u32(data, s + 16)
        raddr = read_u32(data, s + 20)
        sections.append((name, raddr, rsize))
    return sections


def block_hashes(blob, block_size):
    """Hash non-sliding block_size windows. 8-byte SHA1 prefix = cheap block ID."""
    return Counter(
        hashlib.sha1(blob[i:i + block_size]).digest()[:8]
        for i in range(0, len(blob), block_size)
    )


def pct(n, total):
    return round(100.0 * n / total, 1) if total else 0.0


def diff_section(old_blob, new_blob, block_size):
    """Multiset block comparison for one section. Buckets sum to max(old,new)."""
    old_c = block_hashes(old_blob, block_size)
    new_c = block_hashes(new_blob, block_size)
    old_total = sum(old_c.values())
    new_total = sum(new_c.values())
    identical = sum((old_c & new_c).values())   # min-intersection: surviving content
    new_only = new_total - identical             # blocks present only in new
    removed_only = old_total - identical         # blocks present only in old
    changed = min(new_only, removed_only)        # paired replacements (modified)
    net_new = new_only - changed                 # genuine additions (growth)
    net_removed = removed_only - changed         # genuine deletions (shrink)
    total = max(old_total, new_total)
    return {
        'old_size': len(old_blob),
        'new_size': len(new_blob),
        'size_delta': len(new_blob) - len(old_blob),
        'old_blocks': old_total,
        'new_blocks': new_total,
        'identical': identical,
        'changed': changed,
        'new': net_new,
        'removed': net_removed,
        'pct_identical': pct(identical, total),
        'pct_changed': pct(changed, total),
        'pct_new': pct(net_new, total),
        'pct_removed': pct(net_removed, total),
    }


def size_delta_pct(d):
    """Magnitude of raw-size change as % of old section size."""
    return abs(d['size_delta']) / d['old_size'] * 100.0 if d['old_size'] else 0.0


def classify_text(d):
    """RECOMPILE / REFACTOR / MAJOR_CHANGE from .text block survival + size drift."""
    same = d['pct_identical']
    delta = size_delta_pct(d)
    if same < REFACTOR_MIN_SAME or delta > MAJOR_MAX_DELTA:
        return 'MAJOR_CHANGE'
    if same >= RECOMPILE_MIN_SAME and delta < RECOMPILE_MAX_DELTA:
        return 'RECOMPILE'
    # 30-70% survived, or high survival with 1-10% size drift: sigs partly hold.
    return 'REFACTOR'


ACTIONS = {
    'RECOMPILE': 'Offsets shifted, sigs likely survive. Patch-day re-sig viable; '
                 'see skill://pcx-patch-day-playbook.',
    'REFACTOR': 'Some sigs survive, many will not. Playbook applies but expect '
                'manual re-anchoring; see skill://pcx-patch-day-playbook.',
    'MAJOR_CHANGE': 'Block survival too low for a patch-day pass. Assume full '
                    're-RE; the playbook is insufficient on its own.',
}


def build_report(old_data, old_secs, new_data, new_secs, want, block_size):
    """Diff every requested section common to either binary; classify .text."""
    old_map = {n: (r, s) for n, r, s in old_secs}
    new_map = {n: (r, s) for n, r, s in new_secs}
    names = [n for n, _, _ in new_secs] + [n for n, _, _ in old_secs if n not in new_map]
    if want:
        names = [n for n in names if n in want]
    sections = []
    classification = None
    for name in names:
        o_r, o_s = old_map.get(name, (0, 0))
        n_r, n_s = new_map.get(name, (0, 0))
        d = diff_section(old_data[o_r:o_r + o_s], new_data[n_r:n_r + n_s], block_size)
        d['name'] = name
        if name == '.text':
            classification = classify_text(d)
            d['class'] = classification
        sections.append(d)
    tot = sum(max(s['old_blocks'], s['new_blocks']) for s in sections)
    same = sum(s['identical'] for s in sections)
    if classification is None:
        # No .text in scope — fall back to overall survival for the verdict.
        overall = pct(same, tot)
        classification = ('RECOMPILE' if overall >= RECOMPILE_MIN_SAME
                          else 'REFACTOR' if overall >= REFACTOR_MIN_SAME
                          else 'MAJOR_CHANGE')
    summary = {
        'pct_identical': pct(same, tot),
        'pct_changed': round(100.0 - pct(same, tot), 1),
        'classification': classification,
        'recommended_action': ACTIONS[classification],
    }
    return {'sections': sections, 'summary': summary}


def fmt_delta(d):
    return f"{'+' if d >= 0 else '-'}{abs(d)}"


def print_report(report, old_name, new_name):
    print(f"OLD: {old_name}")
    print(f"NEW: {new_name}")
    print()
    hdr = (f"{'SECTION':<10}{'OLD':>11}{'NEW':>11}{'DELTA':>11}"
           f"{'%SAME':>8}{'%CHG':>8}{'%NEW':>8}{'%DEL':>8}  CLASS")
    print(hdr)
    print('-' * len(hdr))
    for s in report['sections']:
        print(f"{s['name']:<10}{s['old_size']:>11}{s['new_size']:>11}"
              f"{fmt_delta(s['size_delta']):>11}"
              f"{s['pct_identical']:>8}{s['pct_changed']:>8}"
              f"{s['pct_new']:>8}{s['pct_removed']:>8}  {s.get('class', '-')}")
    sm = report['summary']
    print()
    print(f"Overall: {sm['pct_identical']}% identical, {sm['pct_changed']}% changed")
    print(f"Verdict: {sm['classification']}")
    print(f"Action:  {sm['recommended_action']}")


def load_binary(path):
    with open(path, 'rb') as f:
        data = f.read()
    secs = get_sections(data)
    if secs is None:
        print(f"Error: {path} is not a valid PE file", file=sys.stderr)
        sys.exit(1)
    return data, secs


def main():
    ap = argparse.ArgumentParser(description='High-level diff summary between two PE versions')
    ap.add_argument('--old', required=True, help='old PE binary')
    ap.add_argument('--new', required=True, help='new PE binary')
    ap.add_argument('--sections', help='comma-separated section names (default: all)')
    ap.add_argument('--block-size', type=int, default=4096, help='block size in bytes (default 4096)')
    ap.add_argument('--json', action='store_true', help='machine-readable output')
    args = ap.parse_args()

    for p in (args.old, args.new):
        if not os.path.isfile(p):
            print(f"Error: {p} not found", file=sys.stderr)
            sys.exit(1)
    if args.block_size < 1:
        print("Error: --block-size must be >= 1", file=sys.stderr)
        sys.exit(1)

    want = {n.strip() for n in args.sections.split(',')} if args.sections else None
    old_data, old_secs = load_binary(args.old)
    new_data, new_secs = load_binary(args.new)
    report = build_report(old_data, old_secs, new_data, new_secs, want, args.block_size)
    if not report['sections']:
        print("Error: no matching sections to compare", file=sys.stderr)
        sys.exit(1)

    if args.json:
        report['old_binary'] = os.path.basename(args.old)
        report['new_binary'] = os.path.basename(args.new)
        report['block_size'] = args.block_size
        print(json.dumps(report, indent=2))
    else:
        print_report(report, os.path.basename(args.old), os.path.basename(args.new))


if __name__ == '__main__':
    main()

File: tools/pe-parser/src/bin/binary-diff-summary.rs
use clap::Parser;
use pe_parser::{load_binary_data, parse_pe_bytes, section_data, Section};
use serde::Serialize;
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;

const RECOMPILE_MIN_SAME: f64 = 70.0;
const RECOMPILE_MAX_DELTA: f64 = 1.0;
const REFACTOR_MIN_SAME: f64 = 30.0;
const MAJOR_MAX_DELTA: f64 = 10.0;

#[derive(Parser)]
#[command(about = "High-level diff summary between two binary versions")]
struct Args {
    #[arg(long)]
    old: PathBuf,
    #[arg(long)]
    new: PathBuf,
    #[arg(long)]
    sections: Option<String>,
    #[arg(long = "block-size", default_value_t = 4096)]
    block_size: usize,
    #[arg(long)]
    json: bool,
}

#[derive(Serialize, Clone)]
struct SectionDiff {
    name: String,
    old_size: usize,
    new_size: usize,
    size_delta: isize,
    old_blocks: usize,
    new_blocks: usize,
    identical: usize,
    changed: usize,
    new: usize,
    removed: usize,
    pct_identical: f64,
    pct_changed: f64,
    pct_new: f64,
    pct_removed: f64,
    #[serde(rename = "class", skip_serializing_if = "Option::is_none")]
    class_name: Option<String>,
}

#[derive(Serialize, Clone)]
struct Summary {
    pct_identical: f64,
    pct_changed: f64,
    classification: String,
    recommended_action: String,
}

#[derive(Serialize)]
struct Report {
    sections: Vec<SectionDiff>,
    summary: Summary,
    #[serde(skip_serializing_if = "Option::is_none")]
    old_binary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    new_binary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    block_size: Option<usize>,
}

fn action_for(classification: &str) -> &'static str {
    match classification {
        "RECOMPILE" => {
            "Offsets shifted, sigs likely survive. Patch-day re-sig viable; see skill://pcx-patch-day-playbook."
        }
        "REFACTOR" => {
            "Some sigs survive, many will not. Playbook applies but expect manual re-anchoring; see skill://pcx-patch-day-playbook."
        }
        _ => {
            "Block survival too low for a patch-day pass. Assume full re-RE; the playbook is insufficient on its own."
        }
    }
}

fn block_counts(blob: &[u8], block_size: usize) -> HashMap<Vec<u8>, usize> {
    let mut out = HashMap::new();
    for chunk in blob.chunks(block_size) {
        *out.entry(chunk.to_vec()).or_insert(0) += 1;
    }
    out
}

fn pct(n: usize, total: usize) -> f64 {
    if total == 0 {
        0.0
    } else {
        ((1000.0 * n as f64 / total as f64).round()) / 10.0
    }
}

fn round1(value: f64) -> f64 {
    (value * 10.0).round() / 10.0
}

fn diff_section(name: String, old_blob: &[u8], new_blob: &[u8], block_size: usize) -> SectionDiff {
    let old_counts = block_counts(old_blob, block_size);
    let new_counts = block_counts(new_blob, block_size);
    let old_blocks: usize = old_counts.values().sum();
    let new_blocks: usize = new_counts.values().sum();
    let mut identical = 0usize;

    for (block, old_count) in &old_counts {
        if let Some(new_count) = new_counts.get(block) {
            identical += (*old_count).min(*new_count);
        }
    }

    let new_only = new_blocks.saturating_sub(identical);
    let removed_only = old_blocks.saturating_sub(identical);
    let changed = new_only.min(removed_only);
    let net_new = new_only - changed;
    let net_removed = removed_only - changed;
    let total = old_blocks.max(new_blocks);

    SectionDiff {
        name,
        old_size: old_blob.len(),
        new_size: new_blob.len(),
        size_delta: new_blob.len() as isize - old_blob.len() as isize,
        old_blocks,
        new_blocks,
        identical,
        changed,
        new: net_new,
        removed: net_removed,
        pct_identical: pct(identical, total),
        pct_changed: pct(changed, total),
        pct_new: pct(net_new, total),
        pct_removed: pct(net_removed, total),
        class_name: None,
    }
}

fn size_delta_pct(diff: &SectionDiff) -> f64 {
    if diff.old_size == 0 {
        0.0
    } else {
        diff.size_delta.unsigned_abs() as f64 / diff.old_size as f64 * 100.0
    }
}

fn classify_text(diff: &SectionDiff) -> String {
    let same = diff.pct_identical;
    let delta = size_delta_pct(diff);
    if same < REFACTOR_MIN_SAME || delta > MAJOR_MAX_DELTA {
        return "MAJOR_CHANGE".to_string();
    }
    if same >= RECOMPILE_MIN_SAME && delta < RECOMPILE_MAX_DELTA {
        return "RECOMPILE".to_string();
    }
    "REFACTOR".to_string()
}

fn load_binary(path: &PathBuf) -> Result<(Vec<u8>, Vec<Section>), String> {
    let data = load_binary_data(path)?;
    let parsed = parse_pe_bytes(&data)?;
    Ok((data, parsed.sections))
}

fn build_report(
    old_data: &[u8],
    old_sections: &[Section],
    new_data: &[u8],
    new_sections: &[Section],
    want: Option<HashSet<String>>,
    block_size: usize,
) -> Report {
    let old_map: HashMap<String, Section> = old_sections
        .iter()
        .map(|sec| (sec.name.clone(), sec.clone()))
        .collect();
    let new_map: HashMap<String, Section> = new_sections
        .iter()
        .map(|sec| (sec.name.clone(), sec.clone()))
        .collect();

    let mut names: Vec<String> = new_sections.iter().map(|sec| sec.name.clone()).collect();
    for sec in old_sections {
        if !new_map.contains_key(&sec.name) {
            names.push(sec.name.clone());
        }
    }
    if let Some(want) = want {
        names.retain(|name| want.contains(name));
    }

    let mut sections = Vec::new();
    let mut classification: Option<String> = None;

    for name in names {
        let old_blob = old_map
            .get(&name)
            .map(|sec| section_data(old_data, sec))
            .unwrap_or(&[]);
        let new_blob = new_map
            .get(&name)
            .map(|sec| section_data(new_data, sec))
            .unwrap_or(&[]);
        let mut diff = diff_section(name.clone(), old_blob, new_blob, block_size);
        if name == ".text" {
            let class_name = classify_text(&diff);
            diff.class_name = Some(class_name.clone());
            classification = Some(class_name);
        }
        sections.push(diff);
    }

    let total: usize = sections
        .iter()
        .map(|sec| sec.old_blocks.max(sec.new_blocks))
        .sum();
    let same: usize = sections.iter().map(|sec| sec.identical).sum();
    let class_name = classification.unwrap_or_else(|| {
        let overall = pct(same, total);
        if overall >= RECOMPILE_MIN_SAME {
            "RECOMPILE".to_string()
        } else if overall >= REFACTOR_MIN_SAME {
            "REFACTOR".to_string()
        } else {
            "MAJOR_CHANGE".to_string()
        }
    });
    let pct_identical = pct(same, total);
    let summary = Summary {
        pct_identical,
        pct_changed: round1(100.0 - pct_identical),
        recommended_action: action_for(&class_name).to_string(),
        classification: class_name,
    };

    Report {
        sections,
        summary,
        old_binary: None,
        new_binary: None,
        block_size: None,
    }
}

fn fmt_delta(delta: isize) -> String {
    if delta >= 0 {
        format!("+{delta}")
    } else {
        format!("-{}", delta.unsigned_abs())
    }
}

fn print_report(report: &Report, old_name: &str, new_name: &str) {
    println!("OLD: {old_name}");
    println!("NEW: {new_name}");
    println!();
    let hdr = format!(
        "{:<10}{:>11}{:>11}{:>11}{:>8}{:>8}{:>8}{:>8}  CLASS",
        "SECTION", "OLD", "NEW", "DELTA", "%SAME", "%CHG", "%NEW", "%DEL"
    );
    println!("{hdr}");
    println!("{}", "-".repeat(hdr.len()));
    for sec in &report.sections {
        println!(
            "{:<10}{:>11}{:>11}{:>11}{:>8.1}{:>8.1}{:>8.1}{:>8.1}  {}",
            sec.name,
            sec.old_size,
            sec.new_size,
            fmt_delta(sec.size_delta),
            sec.pct_identical,
            sec.pct_changed,
            sec.pct_new,
            sec.pct_removed,
            sec.class_name.as_deref().unwrap_or("-")
        );
    }
    println!();
    println!(
        "Overall: {:.1}% identical, {:.1}% changed",
        report.summary.pct_identical, report.summary.pct_changed
    );
    println!("Verdict: {}", report.summary.classification);
    println!("Action:  {}", report.summary.recommended_action);
}

fn basename(path: &PathBuf) -> String {
    path.file_name()
        .and_then(|name| name.to_str())
        .unwrap_or_else(|| path.to_str().unwrap_or(""))
        .to_string()
}

fn main() {
    let args = Args::parse();
    if args.block_size == 0 {
        eprintln!("Error: --block-size must be >= 1");
        std::process::exit(1);
    }

    let (old_data, old_sections) = match load_binary(&args.old) {
        Ok(value) => value,
        Err(e) => {
            eprintln!("Error: {}: {e}", args.old.display());
            std::process::exit(1);
        }
    };
    let (new_data, new_sections) = match load_binary(&args.new) {
        Ok(value) => value,
        Err(e) => {
            eprintln!("Error: {}: {e}", args.new.display());
            std::process::exit(1);
        }
    };
    let want = args.sections.as_ref().map(|sections| {
        sections
            .split(',')
            .map(|name| name.trim().to_string())
            .filter(|name| !name.is_empty())
            .collect::<HashSet<_>>()
    });
    let mut report = build_report(
        &old_data,
        &old_sections,
        &new_data,
        &new_sections,
        want,
        args.block_size,
    );
    if report.sections.is_empty() {
        eprintln!("Error: no matching sections to compare");
        std::process::exit(1);
    }

    let old_name = basename(&args.old);
    let new_name = basename(&args.new);
    if args.json {
        report.old_binary = Some(old_name);
        report.new_binary = Some(new_name);
        report.block_size = Some(args.block_size);
        println!("{}", serde_json::to_string_pretty(&report).unwrap());
        return;
    }
    print_report(&report, &old_name, &new_name);
}
